### PR TITLE
Optimize JsonSerializable and miscellaneous logging fixes

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Accidentally stumbled upon the `JsonSerializable` in Cosmos implementation and noticed that there was redundant `JsonNode` inspections and retrievals. Added a method that merges the has, non-null, and mapping into one `JsonNode` lookup to reduce the number of lookups from 3 to 1.

Additionally, noticed some cases where logging and string formatting had incorrect number of parameters or were using incorrect formatting placeholders. Some cases I fixed as they we simple, for others I left `TODO` comments.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java [hunk 1]
```diff
@@ -100,7 +100,7 @@ public CosmosDiagnosticsContext getDiagnosticsContext() {
     }
 
     void setDiagnosticsContext(CosmosDiagnosticsContext ctx) {
-        checkNotNull("ctx", "Argument 'ctx' must not be null.");
+        checkNotNull(ctx, "Argument 'ctx' must not be null.");
```
Comment (alzimmermsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java:103]: Wasn't checking the input, was instead validating a string constant.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientSideRequestStatistics.java [hunk 2]
```diff
@@ -428,9 +428,7 @@ private void mergeFailedReplica(Set<URI> other) {
             return;
         }
 
-        for (URI uri : other) {
-            this.failedReplicas.add(uri);
-        }
+        this.failedReplicas.addAll(other);
```
Comment (alzimmermsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientSideRequestStatistics.java:431]: `addAll` can be more performant than a loop on `add`

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 3]
```diff
@@ -387,12 +387,12 @@ public static String getNonIdempotentWriteRetryPolicy() {
     public static int getDefaultHttpPoolSize() {
         String valueFromSystemProperty = System.getProperty(HTTP_DEFAULT_CONNECTION_POOL_SIZE);
         if (valueFromSystemProperty != null && !valueFromSystemProperty.isEmpty()) {
-            return Integer.valueOf(valueFromSystemProperty);
+            return Integer.parseInt(valueFromSystemProperty);
```
Comment (alzimmermsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:390]: The boxed parsing methods were being used when returning a primitive. The boxed methods should be used sparingly as they don't guard against nullness inputs (except for Boolean) and most the implementations of the boxed parse methods are wrappers around the primitive parsing methods.

**Comment by alzimmermsft** [comment]
@kushagraThapar this contains a few miscellaneous fixes for the Cosmos libraries, would it possible to get this reviewed?

**Comment by kushagraThapar** [comment]
> @kushagraThapar this contains a few miscellaneous fixes for the Cosmos libraries, would it possible to get this reviewed?

Ack on this, I will review it. @jeet1995 @xinlian12 can you please take a look as well, thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java:103]: great catch, thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientSideRequestStatistics.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientSideRequestStatistics.java:431]: agreed, thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 3 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:390]: Agreed, thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 4]
```diff
@@ -479,6 +479,7 @@ private RxDocumentClientImpl(URI serviceEndpoint,
         this.sessionRetryOptions = sessionRetryOptions;
         this.defaultCustomSerializer = defaultCustomSerializer;
 
+        // TODO: too many arguments, should something additional be added to the log message?
         logger.info(
             "Initializing DocumentClient [{}] with"
                 + " serviceEndpoint [{}], connectionPolicy [{}], consistencyLevel [{}]",
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:485]: protocol doesn't need to be logged; we can remove it from the logger call.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/pkversion/PartitionSynchronizerImpl.java [hunk 5]
```diff
@@ -121,6 +121,7 @@ public Flux<Lease> splitPartition(Lease lease) {
                 return this.leaseManager.createLeaseIfNotExist(addedRangeId, lastContinuationToken, lease.getProperties());
             }, this.degreeOfParallelism)
             .map(newLease -> {
+                // TODO: Too many arguments to log, does something need to be added or removed?
                 logger.info("Partition {} split into new partition and continuation token {}.", leaseToken, newLease.getLeaseToken(), lastContinuationToken);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/pkversion/PartitionSynchronizerImpl.java:125]:
```suggestion
                logger.info("Partition {} split into new partition {} and continuation token {}.", leaseToken, newLease.getLeaseToken(), lastContinuationToken);
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolver.java [hunk 6]
```diff
@@ -272,6 +272,7 @@ private Mono<Utils.ValueHolder<ResolutionResult>> tryResolveServerPartitionAsync
             return addressesObs.flatMap(addressesValueHolder -> {
 
                 if (addressesValueHolder.v == null) {
+                    // TODO: Too few arguments for the log string
                     logger.info(
                         "Could not resolve addresses for identity {}/{}. Potentially collection cache or routing map cache is outdated. Return empty - upper logic will refresh and retry. ",
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolver.java:277]:
```suggestion
                        "Could not resolve addresses for identity {}. Potentially collection cache or routing map cache is outdated. Return empty - upper logic will refresh and retry. ",
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/OpenConnectionAndInitCachesRetryPolicy.java [hunk 7]
```diff
@@ -54,6 +54,7 @@ public Mono<ShouldRetryResult> shouldRetry(Exception e) {
                     Exceptions.isSubStatusCode(clientException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT)) {
 
                 if (this.queryPlanAddressRefreshCount.getAndIncrement() > MAX_ADDRESS_RETRY_COUNT) {
+                    // TODO: logging with exception doesn't use formatting, this should be changed slightly
                     logger.warn("Received {} after exhausted all retry.", e);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/OpenConnectionAndInitCachesRetryPolicy.java:58]:
```suggestion
                    logger.warn("Received exception after exhausting all retries.", e);
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/feedranges/FeedRangeEpkImpl.java [hunk 8]
```diff
@@ -199,6 +199,7 @@ public Mono<RxDocumentServiceRequest> populateFeedRangeFilteringHeaders(
                             // we need to bubble that up to the higher layers to update their
                             // datastructures
                             GoneException goneException = new GoneException(
+                                // TODO: too many arguments for this string.format
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/feedranges/FeedRangeEpkImpl.java:202]: We should remove the last this.range

**Review by kushagraThapar — APPROVED** [review]
This LGTM thanks a lot @alzimmermsft great changes. 

@FabianMeiswinkel / @xinlian12  / @jeet1995 please take a quick look if possible, otherwise, good to merge.

**Comment by kushagraThapar** [comment]
@alzimmermsft once all the PR review feedback is in, worth running the live tests.

**Review by jeet1995 — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java [hunk 9]
```diff
@@ -583,16 +574,13 @@ public ObjectNode getObject(String propertyName) {
...
      */
     Collection<ObjectNode> getCollection(String propertyName) {
-        Collection<ObjectNode> result = null;
-        if (this.propertyBag.has(propertyName) && this.propertyBag.hasNonNull(propertyName)) {
-            result = new ArrayList<ObjectNode>();
-
-            for (JsonNode n : this.propertyBag.findValues(propertyName)) {
-                result.add((ObjectNode) n);
+        return getWithMapping(propertyName, n -> {
+            Collection<ObjectNode> result = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java:578]: Unfortunately, `JsonNode` implements `Iterable` and not `Collection`, so `addAll` can't be used here. Also checked, doesn't seem `ObjectNode` has a good way to get all property nodes.

[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java]: Turns out nothing was using this method, so I deleted it.

**Pull request merged**